src/hdf5_hl.ts

```diff
@@ -685,6 +707,40 @@ export class Dataset extends HasAttrs {
     }
     return processed;
   }
+
+  to_array() {
+    const { value, metadata } = this;
+    const { shape } = metadata;
+    if (!isIterable(value) || typeof value === "string") {
+      return value;
+    }
+    let nested =  create_nested_array(value, shape);
+    return nested;
+  }
+}
+
+function create_nested_array(value: TypedArray | OutputData[], shape: number[]) {
+  // check that shapes match:
+  const total_length = value.length;
+  const dims_product = shape.reduce((previous, current) => (previous * current), 1);
+  if (total_length !== dims_product) {
+    throw new Error(`shape product: ${dims_product} does not match length of flattened array: ${total_length}`);
+  }
+
+  // Get reshaped output:
+  let output = value;
+  const subdims = shape.slice(1).reverse();
+  for (let dim of subdims) {
+    // in each pass, replace input with array of slices of input
+    const new_output = [];
+    const { length } = output;
+    let cursor = 0;
+    while (cursor < length) {
+      new_output.push(output.slice(cursor, cursor += dim));
```

Comment:
Following on my comment about arrays instead of typed arrays:
```suggestion
      new_output.push([...output.slice(cursor, cursor += dim)]);
```

Reply:
Your suggested change would work for datasets that are simple types (float64, int32, etc) but not for Compound datasets, which could still contain TypedArray elements etc.

Also ```tolist``` does not convert to JSON-compatible types automatically, so for the moment ```to_array``` is actually more consistent with ```tolist``` than it would be if we converted the types at the lowest level, but I tend to agree with you in that I usually wish numpy did in fact convert to pure python types when it is called.

Reply:
I see.

Oh well, it seems to work as I expect for boolean arrays so I don't ask more :smile: